compiler/src/dmd/timetrace.d

```diff
@@ -0,0 +1,461 @@
+import dmd.dsymbol;
+import dmd.expression;
+import dmd.root.array;
+import dmd.common.outbuffer;
+import dmd.root.string : toDString;
+
+// Thread local profiler instance (multithread currently not supported because compiler is single-threaded)
+TimeTraceProfiler* timeTraceProfiler = null;
+
+/**
+ * Initialize time tracing functionality.
+ *
+ * Must be called before any other calls to timeTrace functions.
+ *
+ * Params:
+ *   timeGranularityUs = minimum event size in microseconds
+ *   processName = name of this executable
+ */
+extern (C++)
+void initializeTimeTrace(uint timeGranularityUs, const(char)* processName)
+{
+    assert(timeTraceProfiler is null, "Double initialization of timeTraceProfiler");
+    timeTraceProfiler = new TimeTraceProfiler(timeGranularityUs, processName);
+}
+
+/**
+ * Cleanup for time tracing functionality.
+ *
+ * After this, no more calls to timeTrace functions can be made.
+ */
+extern (C++)
+void deinitializeTimeTrace()
+{
+    if (timeTraceProfilerEnabled())
+    {
+        object.destroy(timeTraceProfiler);
+        timeTraceProfiler = null;
+    }
+}
+
+/**
+ * Returns: Whether time tracing is enabled.
+ */
+pragma(inline, true)
+extern (C++)
+bool timeTraceProfilerEnabled()
+{
+    version (LDC)
+    {
+        import ldc.intrinsics : llvm_expect;
+
+        return llvm_expect(timeTraceProfiler !is null, false);
+    }
+    else
+    {
+        return timeTraceProfiler !is null;
+    }
+}
+
+/**
+ * Write all time tracing results so far to JSON, in the Chrome Trace Event Format.
+ * Params:
+ *   buf = output buffer to write JSON into
+ */
+extern (C++)
+void writeTimeTraceProfile(OutBuffer* buf)
+{
+    timeTraceProfiler.writeToBuffer(*buf);
+}
+
+/**
+ * Begin a time trace event.
+ *
+ * Params:
+ *   name_ptr = event name, visible in high level profile view
+ *   detail_ptr = further details, visible when this event is selected
+ *   loc = source location corresponding to this event
+ */
+extern (C++)
+void timeTraceProfilerBegin(scope const(char)* name_ptr, scope const(char)* detail_ptr, Loc loc)
+{
+    import dmd.root.rmem : xarraydup;
+
+    assert(timeTraceProfiler);
+
+    timeTraceProfiler.beginScope(
+        xarraydup(name_ptr.toDString()),
+        xarraydup(detail_ptr.toDString()), loc
+    );
+}
+
+/**
+ * End a time trace event.
+ */
+extern (C++)
+void timeTraceProfilerEnd()
+{
+    assert(timeTraceProfiler);
+    timeTraceProfiler.endScope(TimeTraceEventType.generic, null, null, Loc.initial);
+}
+
+/// Start a new time trace event
+void timeTraceBeginEvent()
+{
+    if (timeTraceProfilerEnabled)
+        timeTraceProfiler.beginScope(null, null, Loc.initial);
+}
+
+/**
+ * End a time tracing event, optionally updating the event name and details
+ * with a delegate.
+ *
+ * Params:
+ *   eventType = what compilation stage the event belongs to
+ *   sym = Dsymbol which was analyzed
+ *   e = Expression which was analyzed
+ */
+void timeTraceEndEvent(TimeTraceEventType eventType)
```

Comment:
`extern(C++)` so it can be used by LDC.

Reply:
overall: think a little about what interface to expose to C++.  Delegates are not possible, but it'd be good if C++ can use the TimeTraceEventType, to have output be nicely uniform across backends, and also provide extra string input for that event.